src/sql/magic.py

```diff
@@ -410,18 +418,57 @@ def interactive_execute_wrapper(**kwargs):
             else:
                 with_ = self._store.infer_dependencies(command.sql_original, args.save)
                 if with_:
-                    command.set_sql_with(with_)
-                    display.message(
-                        f"Generating CTE with stored snippets: {pretty_print(with_)}"
-                    )
+                    query_type = get_query_type(command.sql_original)
+
+                    if query_type != "SELECT":
+                        display.message_warning(
+                            f"Your query is using the following snippets: \
+{', '.join(with_)}. {IF_NOT_SELECT_MESSAGE} CTE generation is disabled"
+                        )
+                    else:
+                        command.set_sql_with(with_)
+                        display.message(
+                            f"Generating CTE with stored snippets: \
+{pretty_print(with_)}"
+                        )
                 else:
                     with_ = None
         else:
+            query_type = get_query_type(command.sql_original)
             if args.with_:
                 raise exceptions.UsageError(
                     "Cannot use --with with CTEs, remove --with and re-run the cell"
                 )
 
+            dependencies = self._store.infer_dependencies(
+                command.sql_original, args.save
+            )
+
+            # The following bit of code checks for dependencies within the parentheses
+            # We do not care about the name outside the parentheses being the same as
+            # another table / snippet here and only check if there are dependencies
+            # in the parentheses.
+            # This is for generating more relevant warnings.
+            parentheses_content = re.findall(
+                r"\((.*)\)", command.sql_original.replace("\n", " ")
+            )
+
+            dependency_in_CTE = []
+            for query in parentheses_content:
+                for dependency in dependencies:
+                    if dependency in query:
+                        dependency_in_CTE.append(dependency)
+
+            if dependency_in_CTE:
+                if query_type != "SELECT":
+                    display_message = IF_NOT_SELECT_MESSAGE
+                else:
+                    display_message = IF_SELECT_MESSAGE
```

Comment:
isn't this branch only running when `is_cte` is true? or maybe I'm missing something. under which cases you fall under one `if` condition or the other?

Reply:
One case is when the CTE is used in a CREATE query like follows:

```
%%sql
WITH my_penguins AS (
    SELECT * FROM penguins.csv
) CREATE TABLE my_penguins_2 AS (
    SELECT * FROM my_penguins
)
```

The query type for this will evaluate to `CREATE`.

A regular query like the following:

```
%%sql
my_penguins AS (
    SELECT * FROM penguins.csv
) SELECT * FROM my_penguins
```

will evaluate to query type of `SELECT`.